pkg/controller/notification/notification_controller_test.go

```diff
@@ -72,22 +88,224 @@ func TestNotificationSentFailure(t *testing.T) {
+	// given
+	ds, mockServer := mockDeliveryService(defaultTemplateLoader())
+
+	mg := mailgun.NewMailgun("crt-test.com", "123")
+	mg.SetAPIBase(mockServer.URL())
+
+	t.Run("test notification delivery ok", func(t *testing.T) {
+		// given
+		userSignup := &v1alpha1.UserSignup{
+			ObjectMeta: newObjectMeta("abc123", "foo@redhat.com"),
+			Spec: v1alpha1.UserSignupSpec{
+				Username:   "foo@redhat.com",
+				GivenName:  "Foo",
+				FamilyName: "Bar",
+				Company:    "Red Hat",
+			},
+		}
+		notification := newNotification("abc123", "test")
+		controller, request, client := newController(t, notification, ds, userSignup)
+
+		// when
+		result, err := controller.Reconcile(request)
+
+		// then
+		require.NoError(t, err)
+		require.True(t, result.Requeue)
+
+		// Load the reconciled notification
+		key := types.NamespacedName{
+			Namespace: operatorNamespace,
+			Name:      notification.Name,
+		}
+		instance := &v1alpha1.Notification{}
+		err = client.Get(context.TODO(), key, instance)
+		require.NoError(t, err)
+
+		test.AssertConditionsMatch(t, instance.Status.Conditions,
+			v1alpha1.Condition{
+				Type:   v1alpha1.NotificationSent,
+				Status: corev1.ConditionTrue,
+				Reason: v1alpha1.NotificationSentReason,
+			},
+		)
+
+		iter := mg.ListEvents(&mailgun.ListEventOptions{Limit: 1})
+		var events []mailgun.Event
+		require.True(t, iter.First(context.Background(), &events))
+		require.True(t, iter.Last(context.Background(), &events))
+		require.Len(t, events, 1)
+		e := events[0]
+		require.IsType(t, &events2.Accepted{}, e)
+		accepted := e.(*events2.Accepted)
+		require.Equal(t, "Foo Bar<foo@redhat.com>", accepted.Recipient)
+		require.Equal(t, "redhat.com", accepted.RecipientDomain)
+		require.Equal(t, "foo", accepted.Message.Headers.Subject)
+		require.Equal(t, "noreply@foo.com", accepted.Message.Headers.From)
+	})
+
+	t.Run("test notification with environment e2e", func(t *testing.T) {
+
+		// given
+		restore := test.SetEnvVarAndRestore(t, "HOST_OPERATOR_ENVIRONMENT", "e2e-tests")
+		defer restore()
+		userSignup := &v1alpha1.UserSignup{
+			ObjectMeta: newObjectMeta("abc123", "jane@redhat.com"),
+			Spec: v1alpha1.UserSignupSpec{
+				Username:   "jane@redhat.com",
+				GivenName:  "jane",
+				FamilyName: "doe",
+				Company:    "Red Hat",
+			},
+		}
+		notification := newNotification("abc123", "test")
+		// pass in nil for deliveryService since send won't be used (sending skipped)
+		controller, request, client := newController(t, notification, nil, userSignup)
+
+		// when
+		result, err := controller.Reconcile(request)
+
+		// then
+		require.NoError(t, err)
+		require.True(t, result.Requeue)
+
+		// Load the reconciled notification
+		key := types.NamespacedName{
+			Namespace: operatorNamespace,
+			Name:      notification.Name,
+		}
+		instance := &v1alpha1.Notification{}
+		err = client.Get(context.TODO(), key, instance)
+		require.NoError(t, err)
+
+		test.AssertConditionsMatch(t, instance.Status.Conditions,
```

Comment:
you could probably refactor this assertion into something like https://github.com/codeready-toolchain/host-operator/blob/master/test/templateupdaterequest/assertion.go#L17-L44 to keep the test shorter. 
(It's just a suggestion, for the sake of readability ;))

Reply:
good idea, changed here: 057eb87e2123a2edfa90a49c007140b845d59010